node: Add reference to mempool in NodeContext

maflcko opened the pull request:
This is the first step toward making the mempool a global that is not initialized before main.

#### Motivation


Currently the mempool is a global that is initialized before the `main` function. This is confusing and easy to get wrong. E.g. the mempool constructor queries state that has not been initialized, like randomness (fixed), or command line arguments (not an issue last time I checked). Also without having the chainstate (chain tip) initialized first, it doesn't make conceptually sense to have a mempool, since the mempool builds txs on top of the utxo set (chain tip).

Finally, in the future someone might want to run a consensus-only full node (`-nowallet -noblockfilter -no... -nomempool` command line options) that only verifies blocks and updates the utxo set.

This is conceptually the same change that has already been done for the connection manager `CConnman`.

ariard commented on src/rpc/blockchain.cpp:
```diff
@@ -518,7 +518,9 @@ static UniValue getrawmempool(const JSONRPCRequest& request)
 
-    return MempoolToJSON(::mempool, fVerbose);
+    const CTxMemPool* tx_pool = MempoolInstance();
+    CHECK_NONFATAL(tx_pool);
```
Due to RPC requests only being served after `SetRPCWarmupFinished` even after mempool is switched to a struct initialized during init sequence, it shouldn't never be null (but better to be secure if we refactor our init logic)

MarcoFalke replied:
Yes, obviously this should be adjusted when the mempool is truly optional. However I don't see the reason to add dead code right now.

ariard replied:
My point was more currently the `CHECK_FATAL` is useless because RPC server accepting requests is the last thing in our init sequence. `CHECK_FATAL` hitting  would be a glaring logic error.

MarcoFalke replied:
> CHECK_FATAL is useless

CHECK_NONFATAL is not useless. Its exact use case it to be used when the condition is assumed to be always true. This is the case here. See https://dev.visucore.com/bitcoin/doxygen/check_8h.html#a46a3e27097aa5e94bbf62075bad7016f

I plan to introduce a `-nomempool` command line option, which makes the mempool go away (optional). When I fail to update this line of code, it will turn into a logic error and will properly throw a bug report.

ariard reviewed (approved):
Code review ACK ae73b97.

Going further in this direction, what would be the next steps ? Passing a mempool pointer to `g_chainstate` ?

MarcoFalke commented:
> Going further in this direction, what would be the next steps ? Passing a mempool pointer to g_chainstate ?

good question. I have decided against that to not interwind chainstate (consensus) with mempool (policy). Otherwise this should be a member function: `CChainState::GetMempool`.

ariard commented:
> I have decided against that to not interwind chainstate (consensus) with mempool (policy)

That means keeping `::mempool` as a global on the long-term? Anyway better to process incrementally, curious of your next PRs on this.

MarcoFalke commented:
> That means keeping ::mempool as a global on the long-term? Anyway better to process incrementally, curious of your next PRs on this.

Yes, but it will become `Optional<CTxMemPool>` (or similar)

ryanofsky commented on src/validation.cpp:
```diff
@@ -132,6 +132,10 @@ CTxMemPool mempool(&feeEstimator);
+CTxMemPool* MempoolInstance()
+{
+    return &::mempool;
+}
```
If this function is supposed to replace uses of ::mempool in existing code, I think it'd be better to return a reference:

```c++
CTxMemPool& MempoolInstance()
{
    return ::mempool;
}
```

now and change it to:

```c++
CTxMemPool& MempoolInstance()
{
    CHECK_NONFATAL(::g_mempool);
    return *::g_mempool;
}
```

in the future when mempool is replaced by an optional or unique_ptr or similar.

Using a reference instead of a pointer for existing code is better because it won't burden the code with checking for nulls.

re: https://github.com/bitcoin/bitcoin/pull/17407/files#issue-338259412

> Finally, in the future someone might want to run a consensus-only full node"

If new code is going to be added or existing code is going to be changed that expects the mempool to be null, this code could probably just access `g_mempool` directly instead of using the helper function. Or you could add different helper functions `CTxMemPool* OptionalMempoolInstance()` or `bool MempoolEnabled()` if referencing the global is a problem.

MarcoFalke replied:
> If this function is supposed to replace uses of ::mempool in existing code, I think it'd be better to return a reference:

It is *not* supposed to replace ::mempool in existing code. It is supposed to be what you refer to as "different helper function". I am happy to rename it to `OptionalMempoolInstance`, but I think that that is a bit too verbose.

ryanofsky replied:
Unclear what the next steps are here. It's hard to see how a function that doesn't log or check anything but just returns a pointer to a global could be useful, but it'd help if you could describe more literally how the function and global and function calls will change in the future.

jnewbery commented:
> Currently the mempool is a global that is initialized before the main function.

This PR doesn't appear to change that, but it sounds like this is the first step in some larger changes. Perhaps you could look to a branch that illustrates what you're planning to do next.

MarcoFalke commented:
Ok, will upload my branches tomorrow

Commit:
node: Add reference to mempool in NodeContext

Currently it is an alias to the global ::mempool and should be used as
follows.

* Node code (validation and transaction relay) can use either ::mempool
  or node.mempool, whichever seems a better fit.
* RPC code should use the added convenience getter EnsureMempool, which
  makes sure the mempool exists before use. This prepares the RPC code
  to a future where the mempool might be disabled at runtime or compile
  time.
* Test code should use m_node.mempool directly, as the mempool is always
  initialized for tests.

Commit:
test: Replace recursive lock with locking annotations

Also, use m_node.mempool instead of the global

Commit:
scripted-diff: Replace ::mempool with m_node.mempool in tests

-BEGIN VERIFY SCRIPT-
 # tx pool member access (mempool followed by dot)
 sed --regexp-extended -i -e 's/(::)?\<mempool\>\.([a-zA-Z])/m_node.mempool->\2/g' $(git grep -l mempool ./src/test)
 # plain global (mempool not preceeded by dot, but followed by comma)
 sed --regexp-extended -i -e 's/([^\.])(::)?\<mempool\>,/\1*m_node.mempool,/g'     $(git grep -l mempool ./src/test)
-END VERIFY SCRIPT-

MarcoFalke commented:
Ok, added some more documentation about future code design in the first commit

ryanofsky commented on src/rpc/protocol.h:
```diff
@@ -63,6 +63,9 @@ enum RPCErrorCode
     RPC_CLIENT_P2P_DISABLED         = -31, //!< No valid connection manager instance found
 
+    //! Chain errors
+    RPC_CLIENT_TX_POOL_DISABLED     = -33, //!< No transaction pool instance found
```
In commit "node: Add reference to mempool in NodeContext" (fa5329abb91ec4876f61967848901c6805ab5503)

Would it be clearer to refer to the mempool instead of the tx pool or transaction pool? (I don't actually know, I'm just used to mempool over transaction pool.)

MarcoFalke replied:
Done. Renamed TX_POOL to MEMPOOL

ryanofsky commented on src/test/miner_tests.cpp:
```diff
@@ -203,6 +197,11 @@ static void TestPackageSelection(const CChainParams& chainparams, const CScript&
 // NOTE: These tests rely on CreateNewBlock doing its own self-validation!
 BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
 {
+    const auto TestSequenceLocks = [this](const CTransaction& tx, int flags) {
```
In commit "test: Replace recursive lock with AssertLockHeld" (fa19ddc2bf8ee8f1ce621611a90f8b9e380b1517)

Instead of `TestSequenceLocks` being a lambda and `TestPackageSelection` in fafe3d88d34f04c798d000ca0b315830d63a80eb taking an unconventionally named `m_node` argument, I think it might be better if both were test fixture methods:

```c++
class MinerTest : public TestingSetup
{
    bool TestSequenceLocks(...);
    void TestPackageSelection(...);
};

bool MinerTest::TestSequenceLocks(...) {...}
void MinerTest::TestPackageSelection(...) {...}

BOOST_FIXTURE_TEST_CASE(CreateNewBlock_validity, MinerTest) { ... }
```

This should make the diffs here smaller and scripted diff simpler because `TestSequenceLocks` wouldn't need to move and `TestPackageSelection` arguments wouldn't have to change.

MarcoFalke replied:
Fixed, I think

ryanofsky reviewed (approved):
Light code review ACK fafe3d88d34f04c798d000ca0b315830d63a80eb (can review more, just let me know)

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -53,6 +54,14 @@ static Mutex cs_blockchange;
 static std::condition_variable cond_blockchange;
 static CUpdatedBlock latestblock;
 
+CTxMemPool& EnsureMemPool()
```
Sounds like a nice follow-up pull request

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -53,6 +54,14 @@ static Mutex cs_blockchange;
 
+CTxMemPool& EnsureMemPool()
+{
+    if (!g_rpc_node->mempool) {
```
Done. added assert

MarcoFalke commented:
Addressed @ryanofsky feedback

jnewbery commented:
Concept ACK. Can you update the PR description, which makes it sound like this PR changes mempool from being a global that is initialized before `main`.

MarcoFalke commented:
Updated OP as requested by @jnewbery

ryanofsky reviewed (approved):
Code review ACK faf24f1e82e77e625bba616412ff5ee3f12e88a5. Just a few suggested changes since last review, constant rename, test / scripted diff cleanup.

It'd be nice to see followup beginning to make use of the new EnsureMempool function. Could be done either here or in a separate PR depending on how complicated it is.

MarcoFalke commented:
> Could be done either here or in a separate PR depending on how complicated it is.

Not complicated, but not a scripted-diff either. Maybe it could be combined with https://github.com/bitcoin/bitcoin/pull/17407#discussion_r346488421, but that sounded more complicated to me.

jnewbery reviewed (commented):
Looks good. Please change 'transaction pool' to mempool in the user facing error message and code comments.

Is there a reason we can't make mempool a unique ptr in `NodeContext`?

MarcoFalke commented:
> Is there a reason we can't make mempool a unique ptr in NodeContext?

Yes, the memory is not yet managed in `NodeContext`. The mempool is created before `main`, and the NodeContext is only created after main.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -53,6 +54,15 @@ static Mutex cs_blockchange;
+{
+    CHECK_NONFATAL(g_rpc_node);
+    if (!g_rpc_node->mempool) {
+        throw JSONRPCError(RPC_CLIENT_MEMPOOL_DISABLED, "Transaction pool disabled or instance not found");
```
Done. Removed "transaction pool"

jnewbery commented:
utACK fa538813b1c382cf135cbf2a0cc3fa01f36964d8

Looking forward to mempool being a unique ptr managed by `NodeContext` in a future PR.

ariard commented on src/rpc/protocol.h:
```diff
@@ -63,6 +63,9 @@ enum RPCErrorCode
     RPC_CLIENT_INVALID_IP_OR_SUBNET = -30, //!< Invalid IP/Subnet
     RPC_CLIENT_P2P_DISABLED         = -31, //!< No valid connection manager instance found
 
+    //! Chain errors
```
nit: I think this change dissociate further chain from mempool, so may qualify better as a mempool error.

MarcoFalke replied:
I think it is fine to keep as is, as mempool and chain are closely related. Also, I'd rather not invalidate ACKs over this.

ariard reviewed (commented):
Tested ACK fa53881.
